service/messageHandler_test.go

```diff
@@ -0,0 +1,163 @@
+func (p *mockMessageProducer) SendMessage(message kafka.FTMessage) error {
+	args := p.Called(message)
+	p.received = append(p.received, message)
+	return args.Error(0)
+}
+
+func TestMessageMapped(t *testing.T) {
+	whitelist := regexp.MustCompile(strings.Replace(testSystemId, ".", `\.`, -1))
+	mp := &mockMessageProducer{}
+	mp.On("SendMessage", mock.AnythingOfType("kafka.FTMessage")).Return(nil)
+
+	service := NewAnnotationMapperService(whitelist, mp)
+
+	contentUuid := uuid.NewV4().String()
+	brandUuid := uuid.NewV4().String()
+	mentionsUuid := uuid.NewV4().String()
+	inbound := kafka.FTMessage{
+		Headers: map[string]string{
+			"Origin-System-Id": testSystemId,
+			"X-Request-Id":     testTxId,
+		},
+		Body: fmt.Sprintf(`{
+		"uuid":"%s",
+		"submittedBy":"test-user",
+		"annotations":[
+		    {
+		        "predicate":"%s",
+		        "id":"%s"
+		    },
+		    {
+		        "predicate":"%s",
+		        "id":"%s"
+		    }
+		]
+		}`, contentUuid,
+			hasBrand, brandUuid,
+			mentions, mentionsUuid),
+	}
+
+	err := service.HandleMessage(inbound)
+	assert.NoError(t, err)
+
+	mp.AssertExpectations(t)
+
+	require.Len(t, mp.received, 1, "messages sent to producer")
+
+	actual := mp.received[0]
+	assert.Equal(t, testTxId, actual.Headers["X-Request-Id"], "transaction_id should be propagated")
+
+	actualBody := ConceptAnnotations{}
+	json.NewDecoder(strings.NewReader(actual.Body)).Decode(&actualBody)
+
+	assert.Equal(t, contentUuid, actualBody.UUID, "content uuid")
+
+	actualAnnotations := actualBody.Annotations
+	assert.Len(t, actualAnnotations, 2, "annotations")
+
+	foundBrand := false
+	foundMentions := false
+	for _, ann := range actualAnnotations {
+		switch ann.Thing.Predicate {
+		case "isClassifiedBy":
+			assert.Equal(t, brandUuid, ann.Thing.ID, "brand annotation")
+			foundBrand = true
+
+		case "mentions":
+			assert.Equal(t, mentionsUuid, ann.Thing.ID, "mentions annotation")
+			foundMentions = true
+		}
+	}
+	assert.True(t, foundBrand, "expected brand predicate was not found")
+	assert.True(t, foundMentions, "expected mentions predicate was not found")
+}
+
+func TestSourceNonMatchingWhitelistIsIgnored(t *testing.T) {
+	whitelist := regexp.MustCompile(`"http://www\.example\.com/ft-system`)
+	mp := &mockMessageProducer{}
+	service := NewAnnotationMapperService(whitelist, mp)
+	inbound := kafka.FTMessage{
+		Headers: map[string]string{"Origin-System-Id": testSystemId},
+		Body:    `{"foo":"bar"}`,
+	}
+
+	actual := service.HandleMessage(inbound)
+	assert.NoError(t, actual)
+	mp.AssertExpectations(t)
+}
+
+func TestSyntacticallyInvalidJsonIsRejected(t *testing.T) {
+	whitelist := regexp.MustCompile(strings.Replace(testSystemId, ".", `\.`, -1))
+	mp := &mockMessageProducer{}
+	service := NewAnnotationMapperService(whitelist, mp)
+	inbound := kafka.FTMessage{
+		Headers: map[string]string{"Origin-System-Id": testSystemId},
+		Body:    `{"foo":"bar"`,
+	}
+
+	actual := service.HandleMessage(inbound)
+	assert.Error(t, actual, "error message")
```

Comment:
Should this check the error message?

Reply:
I'd also leave the `"error message"` text out of the assertion, `assert.Error` is explanation enough of what is happening in that line

Reply:
As the app itself isn't in control of the error message here, it's sensitive to changes in Go, so I'd rather not make an assertion about the text of the error message. Fair enough about the final parameter, though.